packages/contracts/contracts/appDefinitions/HighRollerApp.sol

```diff
@@ -0,0 +1,182 @@
+      uint256 playerSecondNumber = state.commitNum;
+
+
+      if (keccak256(abi.encodePacked(salt, playerFirstNumber)) == state.commitHash) {
+        nextState.winner = chooseWinner(playerFirstNumber, playerSecondNumber);
+      } 
+      else {
+        nextState.winner = Player.SECOND;
+      }
+    } else {
+      revert("Invalid action type");
+    }
+    return abi.encode(nextState);
+  }
+
+  function resolve(AppState state, Transfer.Terms terms)
+    public
+    pure
+    returns (Transfer.Transaction)
+  {
+    uint256[] memory amounts = new uint256[](1);
+    amounts[0] = terms.limit;
+
+    address[] memory to = new address[](1);
+    uint256 player;
+
+    // TODO Need to handle TIE winner state
+    if (state.stage == Stage.DONE) {
+      player = uint256(state.winner);
+    } else {
+      // The player who is not the turn taker
+      player = 1 - uint256(getTurnTaker(state));
+    }
+    to[0] = state.playerAddrs[player];
+
+    bytes[] memory data = new bytes[](2);
+
+    return Transfer.Transaction(
+      terms.assetType,
+      terms.token,
+      to,
+      amounts,
+      data
+    );
+  }
+
+  function chooseWinner(uint256 num1, uint256 num2) 
+    public // Does it matter if this is public or private? The client can poll this if they want?
+    pure
+    returns (Player)
+  {
+    bytes32 finalHash = calculateFinalHash(num1, num2);
+    (bytes8 hash1, bytes8 hash2, bytes8 hash3, bytes8 hash4) = split32Hashto8(finalHash);
+    uint dice1 = bytes8toDiceRoll(hash1);
+    uint dice2 = bytes8toDiceRoll(hash2);
+    uint dice3 = bytes8toDiceRoll(hash3);
+    uint dice4 = bytes8toDiceRoll(hash4);
+    uint256 total1 = dice1 + dice2;
+    uint256 total2 = dice3 + dice4;
+    if(total1 > total2) {
+        return Player.FIRST;
+    }
+    else if(total1 < total2) {
+        return Player.SECOND;
+    }
+    return Player.TIE;
+  }
+
+  function calculateFinalHash(uint256 num1, uint256 num2) 
+    public // Does it matter if this is public or private? The client can poll this if they want?
+    pure
+    returns (bytes32)
+  {
+    uint256 mult = num1 * num2;
+    return keccak256(abi.encodePacked(mult));
+  }
+
+  function split32Hashto8(bytes32 finalHash) 
+    public // Does it matter if this is public or private? The client can poll this if they want?
+    pure
+    returns (bytes8 dice1, bytes8 dice2, bytes8 dice3, bytes8 dice4)
+  {
+    bytes32 mask1 = 0xffffffffffffffff000000000000000000000000000000000000000000000000;
+    bytes32 mask2 = 0x0000000000000000ffffffffffffffff00000000000000000000000000000000;
+    bytes32 mask3 = 0x00000000000000000000000000000000ffffffffffffffff0000000000000000;
+    bytes32 mask4 = 0x000000000000000000000000000000000000000000000000ffffffffffffffff;
+    bytes32 shifted2 = bytes32(uint256(finalHash & mask2) * 2 ** (4*16));
+    bytes32 shifted3 = bytes32(uint256(finalHash & mask3) * 2 ** (4*32));
+    bytes32 shifted4 = bytes32(uint256(finalHash & mask4) * 2 ** (4*48));
```

Comment:
Whoa. Why aren't you using bitshifts? Why do you have to predefine the masks at certain offsets? Does Solidity not play nice with bit trickery?

Reply:
I don't remember where I read but bitshifting using `>>` in Solidity can sometimes be very expensive in terms of gas. 
I could convert this to normal bit shifting if thats whats wanted.

Reply:
Really? That surprises me. In the processor they're the exact same operation AFAIK.

If that's the case though, definitely optimize for gas cost over code readability.

Reply:
I will try to find where I read about it. Googling away

Reply:
@Alonski Did you end up finding proof of higher gas cost?

Reply:
Here's a cleaner implementation you can use

```solidity
function cutBytes32(bytes32 h)
    public
    pure
    returns (bytes8 q1, bytes8 q2, bytes8 q3, bytes8 q4)
{
    assembly {
        let freemem_pointer := mload(0x40)
        mstore(add(freemem_pointer, 0x00), h)
        q1 := mload(add(freemem_pointer, 0x00))
        q2 := mload(add(freemem_pointer, 0x08))
        q3 := mload(add(freemem_pointer, 0x10))
        q4 := mload(add(freemem_pointer, 0x18))
    }
}
```

Reply:
@mvanderh I am going after @snario recommendation anyways now. :)